staker/multi_protocol_staker.go

```diff
@@ -0,0 +1,207 @@
+package staker
+
+import (
+	"context"
+	"time"
+
+	"github.com/OffchainLabs/bold/solgen/go/bridgegen"
+	boldrollup "github.com/OffchainLabs/bold/solgen/go/rollupgen"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/staker/txbuilder"
+	"github.com/offchainlabs/nitro/util/headerreader"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+)
+
+var assertionCreatedId common.Hash
+
+func init() {
+	rollupAbi, err := boldrollup.RollupCoreMetaData.GetAbi()
+	if err != nil {
+		panic(err)
+	}
+	assertionCreatedEvent, ok := rollupAbi.Events["AssertionCreated"]
+	if !ok {
+		panic("RollupCore ABI missing AssertionCreated event")
+	}
+	assertionCreatedId = assertionCreatedEvent.ID
+}
+
+type MultiProtocolStaker struct {
+	stopwaiter.StopWaiter
+	bridge     *bridgegen.IBridge
+	oldStaker  *Staker
+	boldStaker *BOLDStaker
+}
+
+func NewMultiProtocolStaker(
+	l1Reader *headerreader.HeaderReader,
+	wallet ValidatorWalletInterface,
+	callOpts bind.CallOpts,
+	config L1ValidatorConfigFetcher,
+	blockValidator *BlockValidator,
+	statelessBlockValidator *StatelessBlockValidator,
+	stakedNotifiers []LatestStakedNotifier,
+	confirmedNotifiers []LatestConfirmedNotifier,
+	validatorUtilsAddress common.Address,
+	bridgeAddress common.Address,
+	fatalErr chan<- error,
+) (*MultiProtocolStaker, error) {
+	oldStaker, err := NewStaker(
+		l1Reader,
+		wallet,
+		callOpts,
+		config,
+		blockValidator,
+		statelessBlockValidator,
+		stakedNotifiers,
+		confirmedNotifiers,
+		validatorUtilsAddress,
+		fatalErr,
+	)
+	if err != nil {
+		return nil, err
+	}
+	bridge, err := bridgegen.NewIBridge(bridgeAddress, oldStaker.client)
+	if err != nil {
+		return nil, err
+	}
+	return &MultiProtocolStaker{
+		oldStaker:  oldStaker,
+		boldStaker: nil,
+		bridge:     bridge,
+	}, nil
+}
+
+func (m *MultiProtocolStaker) IsWhitelisted(ctx context.Context) (bool, error) {
+	return m.oldStaker.IsWhitelisted(ctx)
+}
+
+func (m *MultiProtocolStaker) Initialize(ctx context.Context) error {
+	boldActive, rollupAddress, err := m.isBoldActive(ctx)
+	if err != nil {
+		return err
+	}
+	if boldActive {
+		log.Info("BOLD protocol is active, initializing BOLD staker")
+		boldStaker, err := m.setupBoldStaker(ctx, rollupAddress)
+		if err != nil {
+			return err
+		}
+		m.boldStaker = boldStaker
+		return m.boldStaker.Initialize(ctx)
+	}
+	log.Info("BOLD protocol not detected on startup, using old staker until upgrade")
+	return m.oldStaker.Initialize(ctx)
```

Comment:
this will only initialize oldStaker if BOLD is not active on startup, which will cause problems later on.
I think the cleanest solution will be to to move all shared subcomponents from old staker to multiProtocolStaker.
In the end result:
* MultiProtocolStaker will have more fields, and any logic from Staker that is still needed for BOLD. e.g. it will hold  stakedNotifiers, ConfirmedNotifiers, ValidatorWalletInterface
* Multi-Staker will have more logic that will be moved to it from oldStaker. E.g: that adds block_validator to StakedNotifiers on NewMultiProtocolStaker, and  calling multi-protocol's Initialize will call Wallet's Initialize
* setupBoldStaker should not use any field internal to oldStaker unless there's a very good reason - it should only use fields from multi-staker.

Reply:
not strictly necessary - but if possible: have subdir/package in staker for staker/legacy (which will hold staker.go, l1validator.go) and and staker/bold which will hold bold_staker, bold_state_provider
we should be able to see that only multi_protocol_staker includes both of these

Reply:
Done.